File: Data Structures & Algorithms/jump-game-ii/submission-1.cpp
class Solution {
public:
    // greedy (two-pointer) solution, O(n) time and O(1) space 
    int jump(vector<int>& nums) {
        int numJumps = 0;
        int l = 0, r = 0;
        while (r < nums.size() - 1) {
            // compute the furthest jump we can make in range [l, r]
            int furthestJump = 0;
            for (int i = l; i < r + 1; i++) {
                furthestJump = max(furthestJump, i + nums[i]);
            }
            // update the next range and increment the jump counter
            l = r + 1;
            r = furthestJump;
            numJumps++;
        }
        return numJumps;
    }
};

File: Data Structures & Algorithms/longest-strictly-increasing-or-strictly-decreasing-subarray/submission-1.py
class Solution:
    # one-pass solution, O(n) time, O(1) space
    def longestMonotonicSubarray(self, nums: List[int]) -> int:
        increase, decrease = 1, 1
        longest = 1

        for i in range(1, len(nums)):
            if nums[i] == nums[i - 1]:
                increase = decrease = 1
            elif nums[i] > nums[i - 1]:
                increase += 1
                decrease = 1
            else:
                increase = 1
                decrease += 1
            longest = max(longest, increase, decrease)

        return longest